packages/insomnia-app/app/network/libcurl-promise.ts

```
@@ -0,0 +1,255 @@
+import { Readable, Writable } from 'stream';
+import { ValueOf } from 'type-fest';
+
+import { describeByteSize } from '../common/misc';
+import { ResponseHeader } from '../models/response';
+import { ResponsePatch } from './network';
+
+// wraps libcurl with a promise taking options, path to response file, and max timeline size
+// returning a response patch, debug timeline and headerArray
+interface CurlRequestOutput {
+  patch: ResponsePatch;
+  debugTimeline: ResponseTimelineEntry[];
+  headerResults: HeaderResult[];
+}
+
+interface CurlOpt {
+  key: CurlSetOptName;
+  value: CurlSetOptValue;
+}
+
+interface CurlRequestOptions {
+  curlOptions: CurlOpt[];
+  responseBodyPath: string;
+  maxTimelineDataSizeKB: number;
+  cancelId: string;
+  requestBodyPath?: string;
+  isMultipart: boolean;
+}
+
+type CurlSetOptParameters = Parameters<Curl['setOpt']>;
+type CurlSetOptName = CurlSetOptParameters[0];
+type CurlSetOptValue = CurlSetOptParameters[1];
+// NOTE: this is a dictionary of functions to close open listeners
+const cancelCurlRequestHandlers = {};
+export const cancelCurlRequest = id => cancelCurlRequestHandlers[id]();
+export const curlRequest = (options: CurlRequestOptions) => new Promise<CurlRequestOutput>(async resolve => {
+  // Create instance and handlers, poke value options in, set up write and debug callbacks, listen for events
+  const { curlOptions, responseBodyPath, requestBodyPath, maxTimelineDataSizeKB, cancelId, isMultipart } = options;
+  const curl = new Curl();
+  let requestFileDescriptor;
+  const responseBodyWriteStream = fs.createWriteStream(responseBodyPath);
+  // cancel request by id map
+  cancelCurlRequestHandlers[cancelId] = () => {
+    if (requestFileDescriptor && responseBodyPath) {
+      closeReadFunction(requestFileDescriptor, isMultipart, requestBodyPath);
+    }
+    curl.close();
+  };
+  // set the string and number options from network.ts
+  curlOptions.forEach(opt => curl.setOpt(opt.key, opt.value));
```

Do we need to do any error handling here like we used to do in `network.ts`, or will that bubble up correctly as-is?

---

Interesting point, curious if there are any options which could cause a useful error out of setOpt I couldn't find any.

---

Sounds reasonable! It's probably not necessary for us to have an explicit try/catch here for this case, mainly just want to make sure that if something throws in this function for some reason and isn't explicitly caught, it does bubbles form of error (regardless if it's called via ipc or not), eg doesn't just silently fail in the UI